Add optical surface physics input and import surface physics data from Geant4

stognini opened the pull request:
This PR adds `inp::SurfacePhysics` to `inp::OpticalPhysics` and imports data from Geant4 accordingly.

The new `SurfacePhysics` untangles the existing Geant4 code design (with high code branching based on surface type/finish for each surface model) into an interactor/kernel-like structure to make it more suitable for GPUs.

The import process has to move away from `ImportData` to ultimately populate `inp::Problem`. To start this process, the `ImportData` is now starting to be populated with `inp` objects.

stognini committed:
Define inp::SurfacePhysics and add it to inp::OpticalPhysics

stognini committed:
Update documentation

stognini committed:
Draft helper class to load inp::SurfacePhysics

stognini committed:
Refactor ImportData -> LegacyImportData to accomodate inp::OpticalPhysics

stognini committed:
Update ROOT dictionary to include OpaqueIds and new inp:: structs

stognini committed:
Refactor ImportData::legacy throughout the codebase

stognini committed:
Update tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stognini commented:
TODO: 
- Create tests with glisur/unified model data
- Correctly verify that inputed data matches minimum requirements for each model
- Add  detector efficiency to Surface Physics?
   ```cpp
   SurfacePhysics
   {
       using DetectionEfficiency = std::map<SurfaceLayer, Grid>;
       ...
       DetectionEfficiency efficiency;
   };
   ```

hhollenb commented on src/celeritas/inp/Surfaces.hh:
```diff
@@ -0,0 +1,221 @@
+ */
+struct ReflectionForm
+{
+    real_type lambertian_roughness{-1};  //! \todo: range?
```
I'll suggest leaving out this parameter for now. I believe this refers to the `kSURFACEROUGHNESS` property which is another characterization of the material's reflectivity (i.e. in comparison of specular lobe / spike / back-scatter / diffuse). The Lambertian distribution itself doesn't take any parameters, and I can't find any discussion of what model this parameter came from or any examples of it being used. I don't think we lose anything by leaving it out for now, and that way we don't have hanging TODOs in the code

stognini replied:
Correct. That ends up being used here:
https://gitlab.cern.ch/geant4/geant4/-/blob/master/source/processes/optical/src/G4OpBoundaryProcess.cc?ref_type=heads#L1098

For this variable I did not check defaults in the manual yet, just seemed reasonable to add it. If no one really uses it, we just never hit that `if` statement. So happy to remove it from the struct/import and revisit this if we need in the future.

amandalund commented on src/celeritas/io/ImportData.hh:
```diff
@@ -106,6 +107,17 @@ struct ImportData
+struct ImportData
+{
+    LegacyImportData legacy;  //!< Temporary container; Will be deprecated
+    inp::OpticalPhysics optical_physics;  //!< \todo: replace by inp::Physics
```
To avoid changing `data` to `data.legacy` all over the code, is there any reason we can't keep the old `ImportData` struct and just put the `inp::OpticalPhysics` directly in there? We've already started replacing some of the old `Import` structs with `inp` there (e..g. the various grid classes).

stognini replied:
I can do that. That was more of an organizational choice---as `inp` objects are pulled out, they start to show up outside `legacy`. But again, we can just make a clear separation in the original `ImportData` and write the todo documentation. No strong preference here. :)

sethrj replied:
Just saw these comments: see below. Maybe we should have a brief meeting to discuss the restructuring. The whole ImportData refactoring is possibly a big jump in the scope of the next step.

sethrj commented on doc/usage/input/physics.rst:
```diff
@@ -10,18 +10,52 @@ Physics
+stepping loop, where surface physics is the most complex part. Therefore, the
+``OpticalPhysics`` input includes optical photon _generation_ processes (such as
+Cherenkov and scintillation) and surface physics information. The latter
+describing how optical photons should interact with it.
```
I wonder if we want separate "optical generation" physics, volumetric physics for optical photons, and surface physics for optical photons...

sethrj commented on src/celeritas/io/detail/ImportDataConverter.cc:
```diff
@@ -33,39 +33,39 @@ ImportDataConverter::ImportDataConverter(UnitSystem usys)
 void ImportDataConverter::operator()(ImportData* data)
 {
-    for (auto& m : data->geo_materials)
+    for (auto& m : data->legacy.geo_materials)
```
Instead of all these changes, just use `LegacyImportData*` as the argument and call it from a new `ImportData*` function. Same for the data trimmer.

sethrj commented on src/celeritas/mat/MaterialParams.cc:
```diff
@@ -62,14 +62,14 @@ MatterState to_matter_state(ImportMaterialState state)
-    CELER_EXPECT(!data.elements.empty());
+    CELER_EXPECT(!data.legacy.geo_materials.empty());
+    CELER_EXPECT(!data.legacy.phys_materials.empty());
+    CELER_EXPECT(!data.legacy.elements.empty());
```
Again, fewer changes by doing
```c++
MaterialParams::from_import(ImportData const& new_data)
{
    auto const& data = new_data.legacy;
    CELER_EXPECT(!data.geo_materials.empty());
    CELER_EXPECT(!data.phys_materials.empty());
    CELER_EXPECT(!data.elements.empty());
```

sethrj commented on src/celeritas/ext/RootInterfaceLinkDef.h:
```diff
@@ -47,17 +47,33 @@
+#pragma link C++ class celeritas::inp::RoughnessModels+;
+#pragma link C++ class celeritas::inp::SmearRoughness+;
+#pragma link C++ class celeritas::inp::SurfacePhysics+;
+#pragma link C++ class celeritas::inp::SurfaceNames+;
```
Do we really want to add all these types to ROOT? I think we should consider abandoning the ROOT output for the time being. Or at least not writing any of the optical stuff to ROOT: we require Geant4 to be loaded to get the surface geometry information regardless.

sethrj reviewed (commented):
Still reviewing this, but when you see massive find-replace changes of the kind `data` -> `data.legacy` all over the code base, we should pause to consider whether it's the right approach, and probably open a separate PR for the refactoring aspect.

I'm not sure if we need ImportData to include the optical physics data, or if we can keep it alongside it for time being, building it in the same fashion that we do the `inp::Model` components.

If we do end up keeping the surface physics data alongside the import data, let's at least separate out the "ImportData is now ImportLegacyData" into a separate PR.

stognini committed:
Convert ReflectionForm parameters to Grid

stognini committed:
Add surface physics data to lar-sphere.gdml; expand GeantImporter test

stognini committed:
Revert legacy import data

sethrj commented on src/celeritas/inp/Surfaces.hh:
```diff
@@ -0,0 +1,250 @@
+    using DetectionEfficiency = std::map<SurfaceLayer, Grid>;
+    //!@}
+
+    SurfaceNames names;
```
We don't need surface names; those are input by the surface params. If the G4 properties here had separate names we could use those, but they don't.

stognini committed:
Verify model requirements for each surface import

stognini committed:
Minor improvements

stognini committed:
Fix missing include

stognini committed:
Merge branch 'develop' into import-surfaces

stognini committed:
geant_geo is now a weak_ptr

sethrj commented on src/celeritas/inp/Surfaces.hh:
```diff
@@ -0,0 +1,251 @@
+ */
+struct SmearRoughness
+{
+    real_type polishness{-1};  //!< [0, 1] where 1 means maximum polishness
```
`polishness` isn't a word in English; `polish` is, but it's confusing (both a verb and a nationality!). It seems like computer graphics has the answer; it's more standard to use "roughness" (0 is fully specular, 1 fully diffuse) so even though it's backward from Geant4 maybe we shoul duse that instead.

sethrj commented on src/celeritas/inp/Surfaces.hh:
```diff
@@ -0,0 +1,251 @@
+    // The sum of these properties must be equal to 1
+    Grid specular_lobe;  //!< [0, 1] probability
+    Grid specular_spike;  //!< [0, 1] probability
+    Grid back_scatter;  //!< [0, 1] probability
```
```suggestion
    Grid backscatter;  //!< [0, 1] probability
```
but maybe could also be specular_backscatter? Look at the original paper to see what options

sethrj commented on src/celeritas/inp/Surfaces.hh:
```diff
@@ -0,0 +1,251 @@
+    explicit operator bool() const
+    {
+        return unity(specular_lobe) && unity(specular_spike)
+               && unity(back_scatter) && unity(diffuse_lobe)
```
These are way too complicated for input. I'm not sure what level of validation we need, but we don't want to be doing summations over all the elements to determine whether they're "assigned". **All `operator bool` should be executable in constant time** i.e., they can't have loops.

Validation of this level should always be where they're used, so we can use this code later when implementing the roughness physics. For now just return `specular_lobe  && specular_spike && ... ;` if all grids are needed.

sethrj commented on src/celeritas/inp/Surfaces.hh:
```diff
@@ -0,0 +1,251 @@
+{
+    std::map<SurfaceLayer, Polished> polished;
+    std::map<SurfaceLayer, SmearRoughness> smear;
+    std::map<SurfaceLayer, GaussianRoughness> gaussian;
```
Instead of these maps of surface layers, since all surfaces should be assigned and we don't know if any/many layers will be present: let's define a typedef
```c++
template<class T>
using SurfaceLayerVector = std::vector<T>;
```
so for the present it's one element per surface ID, indexed by surface ID; then when we add layers we can easily do
```c++
template<class T>
using SurfaceLayerVector = std::vector<std::variant<T, std::vector<T>>;
```
so that for a given surface we can assign a layer of data but by default it's just a single element.

sethrj replied:
Actually I forgot that each of these is going to be sparse, and since the models should build this in sequential/reproducible order we *should* keep it sorted as map. We also have to think about how we'd handle multiple layers: would we need a separate vector at the top level to keep track of how many interfaces per surface there are, so we can correctly search for the entries in the model maps? Or do we build those on the fly when we construct the surface physics?

sethrj commented on src/celeritas/inp/Surfaces.hh:
```diff
@@ -0,0 +1,251 @@
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/inp/Surfaces.hh
```
Rename `SurfacePhysics` ; it's not how surfaces are defined (geometry) but physics (what happens on them)

sethrj commented on src/celeritas/inp/Surfaces.hh:
```diff
@@ -0,0 +1,251 @@
+    ReflectivityModels reflectivity;
+    RoughnessModels roughness;
+    InteractionModels interaction;
+    DetectionEfficiency efficiency;  //!< \todo: Keep as optional?
```
I think efficiency is a terrible name for what this does in Geant4: from what I can tell, if you define it at all, it immediately absorbs the track without any other physics being applied. I almost think it should be part of a separate "surface detector" group in scoring that forces the surface to become a "detector surface" and then with a given probability passes the energy on to the detector. @hhollenb what do you think?

hhollenb replied:
I believe efficiency refers to the quantum efficiency of the detecting surface. I've definitely heard it used regularly when people talk about SiPMs so I think it's readily understood by people doing instrumentation.

I personally think that detection efficiency should be a property of the sensitive detector associated with that surface / volume. It receives hits that are absorbed photons and can apply whatever quantum efficiency it wants, but that's the responsibility of scoring / digitization and not surface physics.

sethrj replied:
Yeah I definitely am familiar with "intrisic detection efficiency", defined by Knoll as "ε = # pulses / number of radiation quanta incident on the detector" which is the input here. The weird side effect is that it forces the surface to be completely absorbent. So I agree, it is a detector property that forces a surface to be strictly a detector with a given efficiency.

sethrj replied:
So @stognini leave this out for now but add a to-do in the docs for `inp::Scoring`

stognini replied:
Yep. I agree that it makes sense for the quantum efficiency to be in sensitive detectors. I'll make a note in the scoring bit for it

sethrj reviewed (commented):
Some preliminary comments on the input; haven't looked at the loading code. Since the `codespell` pre-commit was added on develop, you can run `pre-commit run -a codespell` to fix the typo(s). Also @pcanal can you help determine why the addition of the new fields (empty for most of the probelms) increased file size by a factor of 2?

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+/*!
+ * Get a string corresponding to the \c G4SurfaceType selection.
+ */
+char const* to_cstring(G4SurfaceType value)
```
These should all be in an anonymous namespace since they're only used within this translation unit. I would also try to use a macro to avoid duplication/typos:
```c++
#define GSPL_SURFACE_PAIR(ENUMVALUE) {G4SurfaceType:: ENUMVALUE, #ENUMVALUE}
```
and write `to_cstring_impl = {GSPL_SURFACE_PAIR(dielectric_metal), ...}`.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+ */
+char const* to_cstring(G4OpticalSurfaceFinish value)
+{
+    static std::map<G4OpticalSurfaceFinish, std::string> const to_cstring_impl = {
```
Also use unordered maps if they don't have to be sorted

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+    inp::SurfacePhysics result;
+
+    for (auto sid : range(SurfaceId(geo_->num_surfaces())))
+    {
```
It might be more natural (and similar to the other builders) to take a `inp::SurfacePhysics&` in the constructor, and call `operator()` for each `SurfaceId` to add it.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+        CELER_ASSERT(g4log_surf);
+        auto* g4surf_prop = g4log_surf->GetSurfaceProperty();
+        CELER_ASSERT(g4surf_prop);
+        auto* g4opt_surf = dynamic_cast<G4OpticalSurface*>(g4surf_prop);
```
Always check a dynamic cast.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+
+        try
+        {
+            result.names.insert({sid, g4opt_surf->GetName()});
```
remove this since we don't need the names (those come from the geometry)

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+                << "failed to convert surface " << g4opt_surf->GetName()
+                << " with model " << to_cstring(g4opt_surf->GetModel())
+                << " and finish " << to_cstring(g4opt_surf->GetFinish())
+                << ": " << e.details().which << ", " << e.details().what);
```
I know I suggested this in https://github.com/celeritas-project/celeritas/issues/1850#issuecomment-3057294846, but what we could also do is `CELER_LOG(error)` here instead of usurping the exception, then just `throw;` to keep the original exception propagating correctly. We could also define a `RichContextException` class and use `std::throw_with_nested`.

sethrj replied:
And we could also wrap this loop over surfaces in a `MultiExceptionHandler` so that we could print errors for all surfaces.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+        CELER_LOG(error) << "inp::ReflectionForm incorrectly set up. Verify "
+                            "that all parameters have the same grid sizes and "
+                            "that their probability sums (for each energy "
+                            "point in the grid) are equal to 1";
```
Should these be error messages or validation failures? And this one in particular, the validation criteria is so complicated that I think we want a greater level of detail, which we should probably do in the physics itself.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+ * Since the total probability for all 4 properties is equal to one, the
+ * diffuse lobe can be calculated by subtracting the other three.
+ */
+inp::Grid GeantSurfacePhysicsLoader::calc_diffuse_lobe(
```
This should be constant or (since it  doesn't even use class data) a free function

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+    for (auto i : range(size))
+    {
+        // diffuse_lobe = 1 - specular_lobe - specular_spike - back_scatter
+        result.y[i] = real_type{1} - sl.y[i] - ss.y[i] - bc.y[i];
```
I don't think we want to have overconstrained input like this. Eq 60 from [Nayar et al](https://doi.org/10.1109/34.85654) has $`I_{im} = C_{dl} + \frac{C_{sl}}{\cos \theta_r} \exp\left(-\frac{\alpha^2}{2\sigma_\alpha^2} \right) + C_{ss} \delta(\theta_i - \theta_r) \delta(\phi_r)`$ it looks like the diffuse lobe is an additional fudge factor, so I think this is the most natural term to drop.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+// PRIVATE MEMBER FUNCTIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
```
```suggestion
//---------------------------------------------------------------------------//
```

There are a couple of these in the PR

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+void GeantSurfacePhysicsLoader::insert_reflectivity(
+    SurfaceId sid,
+    G4OpticalSurface const& surf,
+    detail::GeantMaterialPropertyGetter& get_property,
```
These arguments get passed a lot. I think a helper class (in a local anonymous namespace) that temporarily stores the optical surface, the surface ID, the property getter, and a reference to `result` would help clear up the code. That kind of refactoring could be easily done inside the translation unit by Copilot (I highly recommend claude 3.7 'thinking').

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+    using G4OSM = G4OpticalSurfaceModel;
+    auto const g4model = surf.GetModel();
+
+    if (g4model == G4OSM::glisur)
```
For enums try to use `switch`.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,435 @@
+        {
+            // Smearing is available
+            inp::SmearRoughness smear{polishness};
+            CELER_ASSERT(smear);
```
Assertions about user input need to be VALIDATE; or we should validate in the model later on. (This is probably a bigger picture question about where and how much to validate user input; especially as we add a python front end.)

hhollenb commented:
I remember having some issues with regenerating the root test data and having larger files. There might be an option set somewhere that outputs extra MC truth data or something? I tried tracking it down in my past branches but couldn't find it -- might want to check your simple-driver.py file and whatnot

stognini committed:
Address some comments:
- Rename Surfaces.hh -> SurfacePhysics.hh
- Add anonymous namespace for helper functions
- Improve to_cstring functions
- Refactor back_scatter -> backscatter
- Remove surface names
- Remove detector quantum efficiency
- Remove trailing //-----//

stognini committed:
Address more comments:
- Refactor GeantSurfacePhysicsLoader so that operator() is called for
  every surface id
- Use detail::GeantSurfacePhysicsHelper to simplify the code

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,469 @@
+           GSPL_ST_PAIR(coated)
+#endif
+        };
+    return to_cstring_impl.find(value)->second.c_str();
```
I also didn't notice: instead of std::string just use `const char*`. Since these are "static" strings rather than local, the pointers to them will still be valid. Also, formatting will be improved if you add a trailing comma, and finally make sure to check for invalid/unknown enums.
```suggestion

    static std::unordered_map<G4SurfaceType, char const*> const names = {
        GSPL_ST_PAIR(dielectric_metal),
        GSPL_ST_PAIR(dielectric_dielectric),
        GSPL_ST_PAIR(dielectric_LUT),
        GSPL_ST_PAIR(dielectric_LUTDAVIS),
        GSPL_ST_PAIR(dielectric_dichroic),
        GSPL_ST_PAIR(firsov),
        GSPL_ST_PAIR(x_ray),
#if G4VERSION_NUMBER >= 1110
        GSPL_ST_PAIR(coated),
#endif
    };
    if (auto iter = names.find(value); iter != names.end())
    {
        return iter->second;
    }
    return "UNKNOWN";
```

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,469 @@
+           GSPL_OSM_PAIR(dichroic)};
+    return to_cstring_impl.find(value)->second.c_str();
+
+#undef GSPL_OSM_PAIR
```
Same comments here

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,469 @@
+    auto const& dl = rf.diffuse_lobe;
+    auto const size = sl.x.size();
+    CELER_ASSERT(ss.x.size() == size && bc.x.size() == size
+                 && dl.x.size() == size);
```
Since these are all user input, these shoudl be validate not assert. Also you're assuming they're on the same grid (what if they have same # points but different values). Maybe this isn't something we can easily validate?

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,469 @@
+    auto const& sl = rf.specular_lobe;
+    auto const& ss = rf.specular_spike;
+    auto const& bc = rf.backscatter;
+    auto const& dl = rf.diffuse_lobe;
```
Not sure if you saw the comment about diffuse lobe; I think you should just delete it entirely, and just assert here that each grid point is nonnegative and <= 1.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,469 @@
+ * Populate surface physics data for a given \c SurfaceId .
+ */
+void GeantSurfacePhysicsLoader::operator()(SurfaceId sid,
+                                           inp::SurfacePhysics& result)
```
Pass result into the constructor since it's "invariant"; just give it surface ID.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,469 @@
+                         << to_cstring(surf.GetModel()) << " and finish "
+                         << to_cstring(surf.GetFinish()) << ": "
+                         << e.details().which << ", " << e.details().what;
+    }
```
I meant to log the error's context and then keep propagating the exception so the code doesn't keep going
```suggestion
        throw;
    }
```

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsHelper.hh:
```diff
@@ -0,0 +1,64 @@
+
+//---------------------------------------------------------------------------//
+#if !CELERITAS_USE_GEANT4
+inline GeantSurfacePhysicsHelper::GeantSurfacePhysicsHelper(SurfaceId)
```
I don't think these are necessary; the class is only ever used within g4-only files.

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.hh:
```diff
@@ -0,0 +1,74 @@
+/*!
+ * Brief class description.
+ */
+class GeantSurfacePhysicsLoader
```
Update docs, and this should probably be a detail class?

sethrj commented on src/celeritas/ext/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,469 @@
+    {
+        this->insert_reflectivity(helper, result);
+        this->insert_roughness(helper, result);
+        this->insert_interaction(helper, result);
```
Maybe we should order these like they're used at runtime: roughness, reflectivity, interaction

amandalund commented:
FWIW, it looks like the root file sizes only increased by ~20% rather than 2x, which looks like it might be from the additional branches created for the optical surface physics. #1885 should bring this back down.

I also ran across this in the ROOT documentation:

> To prevent a data member from being written to the file, insert a ! as the first character in a single-line comment (//) following the declaration of the member.
>
> To accommodate doxygen-style documentation, this annotation can also be written as ///<!.

So apparently if we did want to e.g. exclude the optical physics date from being written to the ROOT file we could just make this change:

```diff
-    inp::OpticalPhysics optical_physics;  //!< \todo: replace by inp::Physics
+    inp::OpticalPhysics optical_physics;  ///<! \todo: replace by inp::Physics
```

sethrj commented:
Neat but ridiculous!

amandalund commented:
Haha, right? An LLM told me like two weeks ago I could disable output that way and I assumed it was hallucinating and gave up asking about ROOT. Then I saw that today in the docs, and... it works 🙃

stognini committed:
Address comments:
- Refactor GeantSurfacePhysicsLoader::operator() to only use SurfaceIds
- Remove diffuse lobe; which reduces validation complexity
- Wrap surface import with MultiExceptionHanlder

stognini committed:
Merge branch 'develop' into import-surfaces

stognini committed:
Move GSPL to detail

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsHelper.cc:
```diff
@@ -0,0 +1,60 @@
+                                             std::string const& name)
+{
+    GeantMaterialPropertyGetter get_property{*mpt_};
+    return get_property(dst, name, {ImportUnits::mev, ImportUnits::unitless});
```
Are all the properties MeV on x, unitless on y?

stognini replied:
Yes. At least for now. Added note.

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,427 @@
+ */
+bool unity(inp::Grid const& grid)
+{
+    return std::any_of(grid.y.begin(), grid.y.end(), [](real_type const& val) {
```
The name isn't right ; and it should be checking all the points not "any". Maybe:
```suggestion
bool is_fractional(inp::Grid const& grid)
{
    return std::all_of(grid.y.begin(), grid.y.end(), [](real_type const& val) {
```

hhollenb replied:
Maybe `is_probability`?

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,427 @@
+            break;
+        default:
+            CELER_LOG(error) << "G4SurfaceType '" << to_cstring(interface_type)
+                             << "' not available";
```
Shouldn't these be runtime errors, not error messages?

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+// SURFACE PHYSICS: interaction mechanisms / reflection models.
+//---------------------------------------------------------------------------//
+/*!
+ * Paramaters used by different reflection mechanisms.
```
```suggestion
 * Parameters used by different reflection mechanisms.
```
and also you can note that the three probabilities must sum to less than one at every grid point; and the remainder is diffuse scattering

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+ * Interaction models for different interface types.
+ *
+ * Existing interface types are dielectrict-dielectric and dielectric-metal.
+ *
+ * \todo Future work may allow for custom interfaces.
```
```suggestion
 * This will be extended to allow user-provided interaction kernels.
```

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+    using DetectionEfficiency = std::map<SurfaceLayer, Grid>;
+    //!@}
+
+    ReflectivityModels reflectivity;
+    RoughnessModels roughness;
+    InteractionModels interaction;
```
Let's order them with the sequence they're called; same in the operator bool.
```suggestion;
    RoughnessModels roughness;
    ReflectivityModels reflectivity
    InteractionModels interaction;
```

stognini committed:
Avoid using soft_equal

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+ */
+struct SmearRoughness
+{
+    real_type roughness{-1};  //!< [0, 1] where 1 means maximum roughness
```
```suggestion
    real_type roughness{-1};  //!< Scale from 0 = specular to 1 = diffuse
```

stognini commented:
Yep, I had problems in the past because of that Doxygen inconsistency and took me awhile to figure out why my dictionaries were failing. And `//!<` actually works, so just this guy `//!` is the problem.

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+/*!
+ * Analytic reflectivity: use Fresnel equations.
+ */
+struct ReflectionAnalytic
```
Perhaps FresnelReflection? Could there be other analytic models?

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+ */
+struct ReflectionGrid
+{
+    Grid grid;
```
```suggestion
    Grid reflectivity;
```

stognini replied:
This and the previous comment made me realize the nomenclature inconsistency. How about this instead, since both grid/analytic are reflectivity models:

```cpp
struct ReflectivityModels
{
    std::map<SurfaceLayer, GridReflection> grid;
    std::map<SurfaceLayer, FresnelReflection> fresnel;
};
```

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+
+//---------------------------------------------------------------------------//
+/*!
+ * Simplest surface normal treatment.
+ */
+struct Polished
```
What about
```suggestion
 * A polished (perfectly smooth) surface, where the facet normal is the macroscopic normal.
 */
struct NoRoughness
```

stognini replied:
I like that! Even though negating roughness seems more work than just polished, it does unify the nomenclature throughout all the possible options.

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+//---------------------------------------------------------------------------//
+/*!
+ * Surface reflectivity index, which can be a user-defined Grid, which is a
+ * function of energy (wavelength), or constant.
```
I think these should be `XReflectivity` since we have `XRoughness`. ReflectionGrid sounds like it's a `Grid ` about reflection.

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,427 @@
+ * Collect reflectivity information from a given optical surface.
+ */
+void GeantSurfacePhysicsLoader::insert_reflectivity(
+    detail::GeantSurfacePhysicsHelper& helper)
```
Since the surface physics loader is in the detail namespace, you don't have to add `detail::` here.

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,427 @@
+        refl_mods.grid.insert({sid, std::move(refl_grid)});
+    }
+    refl_mods.analytic.insert({sid, inp::ReflectionAnalytic{}});
+    result_.reflectivity = std::move(refl_mods);
```
This is wrong! You're replacing all the models so you'll only ever end up with the last surface added. Instead, use `auto& refl_mods = result_.reflectivity;` at the start of the function.

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,427 @@
+        // Insert any model that includes user-defined grid reflectivity
+        inp::ReflectionGrid refl_grid;
+        helper.get_property(&refl_grid.grid, "REFLECTIVITY");
+        refl_mods.grid.insert({sid, std::move(refl_grid)});
```
You could define a templated method for your helper class to help
```c++

template<class T>
void GeantSurfacePhysicsHelper::insert(std::map<SurfaceLayer, T>& models, T&& surface_model)
{
    auto&& [iter, inserted] = models.insert({this->surface_id(), std::forward<T>(surface_model)});
    CELER_VALIDATE(inserted,
                   << "duplicate model inserted for surface ID " << this->surface_id());
}
```
so this would become the safer:
```suggestion
        helper.insert(refl_mods.grid, std::move(refl_grid));
```

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,427 @@
+ * Currently, only the Unified model with [polished/ground]backpainted undergo
+ * uniquely through analytical reflection (i.e. Fresnel equations).
+ */
+bool GeantSurfacePhysicsLoader::analytic_reflection_only(
```
This can be a free function

stognini commented:
Still about the ROOT files: After reorganization and what is likely the final format of `SurfacePhysics`, the `lar-sphere.root` goes from 33k to 35k, which seems reasonable for the extra surface info + overhead.

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,427 @@
+        case G4OSM::glisur: {
+            // Get GLISUR surface polish
+            real_type roughness = real_type{1} - surf.GetPolish();
+            if (roughness == soft_equal(real_type{0}, roughness))
```
```suggestion
            if (soft_zero(roughness))
```

sethrj reviewed (commented):
*please* next time submit a shorter pull request. We should've started with a single model and built up, not thrown *everything* all at once.

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,426 @@
+            {
+                // Smearing is available (surf.GetFinish() == G4OSF::ground)
+                // Celeritas' roughness is the complement of Geant4 polish
+                inp::SmearRoughness smear{real_type{1} - surf.GetPolish()};
```
Is  a polish of 1 used in practice? Should we replace it with a polished model?

stognini replied:
Actually, re-re-reading the code, having `ground` finish and `polish = 1.0` is still possible. So you can have a perfectly polished surface that undergoes spike reflection (with finish == polished), or a ground finish that is also perfectly polished (with `GetPolish()` returning 1), but undergoing lobe reflection...

sethrj replied:
Aren't spike vs lobe the reflection model, not the surface facet distribution?

stognini replied:
Correct. But when Geant's `G4OpBoundaryProcessStatus` is set as `LobeReflection` (for `ground` finish)  it also calls `GetFacetNormal()`, which uses `GetPolish()`. So the finish determines if it uses spike or lobe, and in turn the lobe model calls `GetFacetNormal()`.

The `GetFacetNormal()` checks `if (polish < 1)` to calculate `facetNormal = normal + (1. - polish) * smear`, otherwise manually sets `facetNormal = normal`, which is an unnecessary if/else statement, since if polish is 1 the extra term is just zero.

pcanal commented:
> Neat but ridiculous!

Marking as transient should really be an attribute but they did not exist 30 years ago and we have not had the effort to properly introduce/use them.  Related (and reducing the urgency of such a move) is that the transientness can also be marked in the dictionary request file (when it is in the xml format)

sethrj commented on src/celeritas/inp/SurfacePhysics.hh:
```diff
@@ -0,0 +1,206 @@
+/*!
+ * Interaction models for different interface types.
+ *
+ * Existing interface types are dielectrict-dielectric and dielectric-metal.
```
```suggestion
 * Existing interface types are dielectric-dielectric and dielectric-metal.
```

sethrj commented on src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc:
```diff
@@ -0,0 +1,426 @@
+            if (gauss)
+            {
+                result_.roughness.gaussian.insert({sid, std::move(gauss)});
+            }
```
What's the `else` case? You need a rougnness model right? I guess this gets checked in the validation? (That seems error prone.)

stognini committed:
WIP: reorg surface physics loader logic

sethrj commented:
> > Neat but ridiculous!
> 
> Marking as transient should really be an attribute but they did not exist 30 years ago and we have not had the effort to properly introduce/use them. Related (and reducing the urgency of such a move) is that the transientness can also be marked in the dictionary request file (when it is in the xml format)

@pcanal The more disruptive aspect is that it's a "surprise": I don't think most people would think a doxygen-like comment or an unusual comment character would trigger a behavior change in *anything*, unless there's a pragma-style comment a la clang-tidy or other source-slurping tools: have you considered deprecating that behavior and using `// root: hide` instead?

sethrj reviewed (commented):
Some probably-stale comments I left in from Thursday; let's try to get a final draft ready today?

stognini committed:
Minor improvements

stognini committed:
Simplify a bit more; improve comments

stognini committed:
Make reflectivity insertion part of the class

pcanal commented:
> The more disruptive aspect is that it's a "surprise":

'Technically' we were there first :)  and doxygen is the one that 'surprisingly' disable a correctly functioning 'thing' :) :)

 > I don't think most people would think a doxygen-like comment or an unusual comment character would trigger a behavior change in anything

Yep, the original choice was indeed probably too succinct :)
 
>  unless there's a pragma-style comment a la clang-tidy or other source-slurping tools: have you considered deprecating that behavior and using // root: hide instead?

Some variations of this has been considered, it is believe that using `attribute` might make the most sense.  The 'hard' part is to finalize a coherent new 'grammar' covering all the options (there is also syntax to express length of array or to specify storage precision of floating numbers).  So it is non-zero effort that has not yet bumbled up on top of other priorities :(

stognini committed:
Add optical surfaces test

stognini committed:
Fix typos; Fix missing include

stognini committed:
Simplify backpainted insertion; Fix test

sethrj committed:
Refactor helper/loader

- Add emplace helper method
- Fix duplicate reflectivity models
- Remove painted/coated for now
- Improve validation vs logging
- Distinguish between not implemented and invalid
- Improve formatting of unordered maps
- Remove extraneous braces

sethrj committed:
constness

sethrj committed:
Improve log message

stognini committed:
Refactor reflection models; update test harness

stognini committed:
Merge branch 'develop' into import-surfaces

stognini committed:
Update RootJsonDumper test

stognini committed:
Explicitly cast `GetPolish` and `GetSigmaAlpha` to real_type

stognini committed:
Refactor SurfaPhysics to use double

hhollenb reviewed (approved):
Looked at the recent changes and looks good!

sethrj committed:
Fix const and add debug log about what's loading

sethrj committed:
Add docs

sethrj committed:
Work around doxygen bug
